<a href="https://colab.research.google.com/github/AliefReefandi/Grokking-Deep-Learning/blob/main/Chapter_03_Introduction_to_Neural_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Chapter 3 — Introduction to Neural Prediction

Chapter ini membahas bagaimana neural network menghasilkan prediksi menggunakan proses **forward propagation**.

Konsep utama:
- Input
- Weight
- Prediction
- Multiple inputs
- Multiple outputs
- Multiple inputs dan outputs
- Neural network bertingkat
- NumPy

## Tujuan Pembelajaran

Setelah mempelajari chapter ini, kita dapat:

1. Memahami hubungan antara input, weight, dan prediction.
2. Membuat neural network sederhana.
3. Menggunakan beberapa input sekaligus.
4. Menghasilkan beberapa output.
5. Memahami neural network dengan beberapa layer.
6. Menggunakan NumPy untuk operasi neural network.

## Teori Utama

Neural network paling sederhana menerima sebuah input dan menggunakan
weight untuk menghasilkan prediction.

Secara sederhana:

Prediction = Input × Weight

Weight berfungsi untuk menentukan seberapa besar pengaruh input terhadap
prediction.

Jika terdapat beberapa input, setiap input dikalikan dengan weight masing-masing
kemudian hasilnya dijumlahkan. Proses ini disebut **weighted sum** atau
**dot product**.

Neural network juga dapat memiliki beberapa output. Selain itu, neural network
dapat disusun dalam beberapa layer, sehingga output dari satu layer dapat
digunakan sebagai input untuk layer berikutnya.

Proses pengiriman data dari input melewati layer-layer hingga menghasilkan
prediction disebut **forward propagation**.

## 2. Konsep Dasar Neural Prediction

Neural network dapat digunakan untuk menghasilkan prediction dari suatu
input.

Pada bentuk paling sederhana, sebuah input dikalikan dengan sebuah weight:

Prediction = Input × Weight

Weight digunakan untuk menentukan seberapa besar pengaruh input terhadap
prediction.

Sebagai contoh, jika input bernilai 8.5 dan weight bernilai 0.1:

Prediction = 8.5 × 0.1
           = 0.85

Semakin kompleks neural network, jumlah input, weight, output, dan layer
yang digunakan juga dapat bertambah.

## 3. Single Input

Pada contoh pertama, neural network hanya menggunakan satu input dan
satu weight.

Data yang digunakan dalam contoh buku adalah data mengenai jumlah toes
(jari kaki) pemain basket.

In [2]:
weight = 0.1

def neural_network(input, weight):
    prediction = input * weight
    return prediction

number_of_toes = [8.5, 9.5, 10, 9]

input = number_of_toes[0]

pred = neural_network(input, weight)

print(pred)

0.8500000000000001


### Penjelasan

Program tersebut melakukan operasi sederhana:

Input = 8.5
Weight = 0.1

Prediction = 8.5 × 0.1
           = 0.85

Contoh ini menunjukkan bentuk paling sederhana dari neural network,
yaitu satu input dikombinasikan dengan satu weight untuk menghasilkan
satu prediction.

## 4. Multiple Inputs

Neural network tidak hanya dapat menggunakan satu input.

Dalam contoh berikut digunakan tiga input:

- toes → jumlah jari kaki
- wlrec → persentase pertandingan yang dimenangkan
- nfans → jumlah penggemar

Setiap input memiliki weight masing-masing.

Prosesnya:

Input 1 × Weight 1
+
Input 2 × Weight 2
+
Input 3 × Weight 3

Hasil akhirnya disebut weighted sum.

In [3]:
def w_sum(a, b):
    assert(len(a) == len(b))

    output = 0

    for i in range(len(a)):
        output += a[i] * b[i]

    return output


weights = [0.1, 0.2, 0]

def neural_network(input, weights):
    pred = w_sum(input, weights)
    return pred


toes = [8.5, 9.5, 9.9, 9.0]
wlrec = [0.65, 0.8, 0.8, 0.9]
nfans = [1.2, 1.3, 0.5, 1.0]

input = [
    toes[0],
    wlrec[0],
    nfans[0]
]

pred = neural_network(input, weights)

print(pred)

0.9800000000000001


### Perhitungan

Prediction dihitung dengan:

(8.5 × 0.1) + (0.65 × 0.2) + (1.2 × 0)

= 0.85 + 0.13 + 0

= 0.98

Jadi prediction yang dihasilkan adalah 0.98.

Fungsi `w_sum()` digunakan untuk melakukan weighted sum secara manual.

## 5. Multiple Outputs

Neural network juga dapat menghasilkan lebih dari satu output.

Jika satu input diberikan kepada beberapa weight, maka setiap weight
dapat menghasilkan nilai output yang berbeda.

Contohnya, satu input dapat digunakan untuk memprediksi tiga nilai
sekaligus.

Untuk melakukan operasi tersebut digunakan fungsi `ele_mul()`.

In [4]:
weights = [0.3, 0.2, 0.9]

def ele_mul(number, vector):
    output = [0, 0, 0]

    assert(len(output) == len(vector))

    for i in range(len(vector)):
        output[i] = number * vector[i]

    return output


def neural_network(input, weights):
    pred = ele_mul(input, weights)
    return pred


wlrec = [0.65, 0.8, 0.8, 0.9]

input = wlrec[0]

pred = neural_network(input, weights)

print(pred)

[0.195, 0.13, 0.5850000000000001]


### Penjelasan

Input yang digunakan adalah:

0.65

Weight:

[0.3, 0.2, 0.9]

Maka:

0.65 × 0.3 = 0.195
0.65 × 0.2 = 0.130
0.65 × 0.9 = 0.585

Sehingga network menghasilkan tiga output:

[0.195, 0.13, 0.585]

## 6. Multiple Inputs dan Multiple Outputs

Neural network dapat menggabungkan multiple inputs dan menghasilkan
multiple outputs secara bersamaan.

Untuk merepresentasikan kondisi tersebut, weight dapat disimpan dalam
bentuk matrix.

Input:

[input1, input2, input3]

Weight:

[
    [w11, w12, w13],
    [w21, w22, w23],
    [w31, w32, w33]
]

Setiap baris weight digunakan untuk menghasilkan satu output.

In [5]:
def vect_mat_mul(vect, matrix):
    assert(len(vect) == len(matrix))

    output = [0, 0, 0]

    for i in range(len(vect)):
        output[i] = w_sum(vect, matrix[i])

    return output

In [6]:
weights = [
    [0.1, 0.1, -0.3],
    [0.1, 0.2, 0.0],
    [0.0, 1.3, 0.1]
]

Fungsi `vect_mat_mul()` melakukan perkalian antara vector input
dengan matrix weight.

Dengan cara ini, neural network dapat menerima beberapa input dan
menghasilkan beberapa output sekaligus.

## 7. Neural Network dengan Multiple Layers

Neural network dapat memiliki lebih dari satu layer.

Pada neural network dua layer:

Input
  ->
Layer 1
   ->
Layer 2
   ->
Prediction

Output dari layer pertama digunakan sebagai input untuk layer kedua.

Artinya, network melakukan prediction berdasarkan prediction sebelumnya.

Konsep ini disebut **predicting on predictions**.

In [7]:
def neural_network(input, weights):
    hid = vect_mat_mul(input, weights[0])
    pred = vect_mat_mul(hid, weights[1])

    return pred

Pada kode tersebut:

1. Input diproses menggunakan `weights[0]`.
2. Hasilnya disimpan sebagai `hid` atau hidden layer.
3. Hidden layer diproses kembali menggunakan `weights[1]`.
4. Hasil akhirnya menjadi prediction.

Dengan demikian, data mengalir dari input menuju hidden layer
kemudian menuju output.

## 8. Implementasi Menggunakan NumPy

Setelah memahami operasi neural network secara manual, operasi tersebut
dapat dibuat lebih sederhana menggunakan library NumPy.

NumPy menyediakan operasi vector dan matrix sehingga perhitungan
weighted sum dapat dilakukan menggunakan fungsi `dot()`.

In [8]:
import numpy as np

weights = np.array([0.1, 0.2, 0])

def neural_network(input, weights):
    pred = input.dot(weights)
    return pred


toes = np.array([8.5, 9.5, 9.9, 9.0])
wlrec = np.array([0.65, 0.8, 0.8, 0.9])
nfans = np.array([1.2, 1.3, 0.5, 1.0])

input = np.array([
    toes[0],
    wlrec[0],
    nfans[0]
])

pred = neural_network(input, weights)

print(pred)

0.9800000000000001


## 9. Forward Propagation

Forward propagation adalah proses ketika data bergerak dari input
melewati layer-layer neural network hingga menghasilkan prediction.

Secara sederhana:

Input
  ->
Weight
  ->
Weighted Sum
  ->
Hidden Layer
  ->
Output
  ->
Prediction

Pada neural network yang memiliki beberapa layer, proses tersebut
dilakukan secara berurutan dari layer pertama sampai layer terakhir.

Chapter ini berfokus pada bagaimana proses prediction tersebut dilakukan.

## Kesimpulan

Chapter 3 menjelaskan dasar bagaimana neural network menghasilkan
prediction.

Pembahasan dimulai dari neural network sederhana dengan satu input
dan satu weight, kemudian berkembang menjadi multiple inputs,
multiple outputs, matrix, dan multiple layers.

Konsep weighted sum, vector, matrix, dan forward propagation menjadi
dasar penting untuk memahami neural network yang lebih kompleks.

Pada chapter berikutnya, pembahasan akan dilanjutkan dari proses
prediction menuju proses **learning**, yaitu bagaimana neural network
dapat memperbaiki weight menggunakan **gradient descent**.